agent/src/storage/bucket.rs

```diff
@@ -20,39 +21,69 @@ impl S3Storage {
         })
     }
 
-    /// Check if the given data blob is present in our data store, by integrity hash.
-    pub async fn data_by_sri(&self, integrity: Integrity) -> ServalResult<ByteStream> {
-        let key = integrity.to_string();
+    /// Check if the given data blob is present in our data store, by integrity hash. Returns a stream.
+    pub async fn data_by_integrity(&self, integrity: &Integrity) -> ServalResult<ByteStream> {
         let object = self
             .client
             .get_object()
             .bucket(&self.bucket)
-            .key(key)
+            .key(encode(&integrity.to_string()))
             .send()
             .await?;
 
         Ok(object.body)
     }
 
+    pub async fn store_by_integrity(
+        &self,
+        integrity: &Integrity,
+        bytes: &[u8],
+    ) -> ServalResult<Integrity> {
+        let body = ByteStream::from(bytes.to_vec());
+        let result = self
+            .client
+            .put_object()
+            .bucket(&self.bucket)
+            .key(encode(&integrity.to_string())) // integrity string is not url-safe!
```

Comment:
It's a real bummer that the AWS client doesn't just take care of this for us automatically.